// kryo/bson/buffer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const missing_dependency_1 = require("../errors/missing-dependency");
const buffer_1 = require("../types/buffer");
// tslint:disable-next-line:variable-name
let Binary = undefined;
function getBinary() {
    if (Binary === undefined) {
        try {
            // tslint:disable-next-line:no-var-requires no-require-imports
            Binary = require("bson").Binary;
        }
        catch (err) {
            throw missing_dependency_1.createMissingDependencyError("bson", "Required to write buffers to BSON.");
        }
    }
    return Binary;
}
function isBinary(val) {
    return val._bsontype === "Binary";
}
function write(type, val) {
    return new (getBinary())(Buffer.from(val));
}
function read(type, input) {
    let result;
    if (isBinary(input)) {
        // TODO: Fix BSON type definitions
        result = input.value(true);
    }
    else {
        result = input;
    }
    const error = type.testError(result);
    if (error !== undefined) {
        throw error;
    }
    return result;
}
function readTrusted(type, input) {
    return input.value(true);
}
exports.readTrusted = readTrusted;
exports.SERIALIZER = {
    typeName: buffer_1.name,
    write,
    read,
    readTrusted,
};
function register(serializer) {
    serializer.register(exports.SERIALIZER);
}
exports.register = register;
