// kryo/bson/union.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const incident_1 = require("incident");
const union_1 = require("../types/union");
function register(serializer) {
    function write(type, val) {
        return serializer.write(type.trustedMatcher(val), val);
    }
    function readWithVariant(type, input) {
        const variant = type.readMatcher(input, serializer);
        if (variant === undefined) {
            throw new incident_1.Incident("UnknownUnionVariant", "Unknown union variant");
        }
        return [serializer.read(variant, input), variant];
    }
    function read(type, input) {
        return readWithVariant(type, input)[0];
    }
    function readTrustedWithVariant(type, input) {
        const variant = type.readTrustedMatcher(input, serializer);
        return [serializer.readTrusted(variant, input), variant];
    }
    function readTrusted(type, input) {
        return readTrustedWithVariant(type, input)[0];
    }
    serializer.register({
        typeName: union_1.name,
        write,
        read,
        readTrusted,
    });
}
exports.register = register;
