// skype-http/types/api-profile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const array_1 = require("kryo/types/array");
const document_1 = require("kryo/types/document");
const json_1 = require("kryo/types/json");
const null_1 = require("kryo/types/null");
const try_union_1 = require("kryo/types/try-union");
const ucs2_string_1 = require("kryo/types/ucs2-string");
const url_1 = require("./url");
function nullable(type) {
    return new try_union_1.TryUnionType({
        variants: [
            new null_1.NullType(),
            type,
        ],
    });
}
exports.$ApiProfile = new document_1.DocumentType({
    properties: {
        firstname: { type: new ucs2_string_1.Ucs2StringType({ maxLength: Infinity }) },
        lastname: { type: nullable(new ucs2_string_1.Ucs2StringType({ maxLength: Infinity })) },
        birthday: { type: new json_1.JsonType() },
        gender: { type: new json_1.JsonType() },
        language: { type: new json_1.JsonType() },
        country: { type: new json_1.JsonType() },
        province: { type: new json_1.JsonType() },
        city: { type: new json_1.JsonType() },
        homepage: { type: new json_1.JsonType() },
        about: { type: new json_1.JsonType() },
        emails: { type: new array_1.ArrayType({ itemType: new ucs2_string_1.Ucs2StringType({ maxLength: Infinity }), maxLength: Infinity }) },
        jobtitle: { type: new json_1.JsonType() },
        phoneMobile: { type: new json_1.JsonType() },
        phoneHome: { type: new json_1.JsonType() },
        phoneOffice: { type: new json_1.JsonType() },
        mood: { type: new json_1.JsonType() },
        richMood: { type: new json_1.JsonType() },
        avatarUrl: { type: nullable(url_1.$Url) },
        username: { type: new ucs2_string_1.Ucs2StringType({ maxLength: Infinity }) },
    },
    ignoreExtraKeys: true,
});
