// kryo/qs/float64.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const invalid_float64_1 = require("../errors/invalid-float64");
const invalid_type_1 = require("../errors/invalid-type");
const float64_1 = require("../types/float64");
function write(type, value) {
    if (isNaN(value)) {
        return "NaN";
    }
    else if (value === Infinity) {
        return "+Infinity";
    }
    else if (value === -Infinity) {
        return "-Infinity";
    }
    return value.toString(10);
}
function read(type, input) {
    if (typeof input !== "string") {
        throw invalid_type_1.createInvalidTypeError("string", input);
    }
    switch (input) {
        case "NaN":
            if (!type.allowNaN) {
                throw invalid_float64_1.createInvalidFloat64Error(input);
            }
            return NaN;
        case "+Infinity":
            if (!type.allowInfinity) {
                throw invalid_float64_1.createInvalidFloat64Error(input);
            }
            return Infinity;
        case "-Infinity":
            if (!type.allowInfinity) {
                throw invalid_float64_1.createInvalidFloat64Error(input);
            }
            return -Infinity;
        default:
            const val = parseFloat(input);
            const error = type.testError(val);
            if (error !== undefined) {
                throw error;
            }
            return val;
    }
}
function readTrusted(type, input) {
    switch (input) {
        case "NaN":
            return NaN;
        case "+Infinity":
            return Infinity;
        case "-Infinity":
            return -Infinity;
        default:
            return parseFloat(input);
    }
}
exports.SERIALIZER = {
    typeName: float64_1.name,
    write,
    read,
    readTrusted,
};
function register(serializer) {
    serializer.register(exports.SERIALIZER);
}
exports.register = register;
