// kryo/types/boolean.mjs
import { createInvalidTypeError } from "../errors/invalid-type";
export const name = "boolean";
export class BooleanType {
    constructor() {
        this.name = name;
    }
    toJSON() {
        /* tslint:disable-next-line:return-undefined */
        return undefined;
    }
    readTrustedJson(input) {
        return input;
    }
    readJson(input) {
        if (typeof input !== "boolean") {
            throw createInvalidTypeError("boolean", input);
        }
        return input;
    }
    writeJson(val) {
        return val;
    }
    testError(val) {
        if (typeof val !== "boolean") {
            return createInvalidTypeError("boolean", val);
        }
        return undefined;
    }
    test(val) {
        return this.testError(val) === undefined;
    }
    equals(val1, val2) {
        return val1 === val2;
    }
    clone(val) {
        return val;
    }
    /**
     * @param oldVal
     * @param newVal
     * @returns `true` if there is a difference, `undefined` otherwise
     */
    diff(oldVal, newVal) {
        /* tslint:disable-next-line:strict-boolean-expressions */
        return (oldVal !== newVal) || undefined;
    }
    patch(oldVal, diff) {
        return oldVal === (diff === undefined);
    }
    reverseDiff(diff) {
        return diff;
    }
    squash(diff1, diff2) {
        /* tslint:disable-next-line:strict-boolean-expressions */
        return (diff1 !== diff2) && undefined;
    }
}
