// kryo/types/boolean.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const invalid_type_1 = require("../errors/invalid-type");
exports.name = "boolean";
class BooleanType {
    constructor() {
        this.name = exports.name;
    }
    toJSON() {
        /* tslint:disable-next-line:return-undefined */
        return undefined;
    }
    readTrustedJson(input) {
        return input;
    }
    readJson(input) {
        if (typeof input !== "boolean") {
            throw invalid_type_1.createInvalidTypeError("boolean", input);
        }
        return input;
    }
    writeJson(val) {
        return val;
    }
    testError(val) {
        if (typeof val !== "boolean") {
            return invalid_type_1.createInvalidTypeError("boolean", val);
        }
        return undefined;
    }
    test(val) {
        return this.testError(val) === undefined;
    }
    equals(val1, val2) {
        return val1 === val2;
    }
    clone(val) {
        return val;
    }
    /**
     * @param oldVal
     * @param newVal
     * @returns `true` if there is a difference, `undefined` otherwise
     */
    diff(oldVal, newVal) {
        /* tslint:disable-next-line:strict-boolean-expressions */
        return (oldVal !== newVal) || undefined;
    }
    patch(oldVal, diff) {
        return oldVal === (diff === undefined);
    }
    reverseDiff(diff) {
        return diff;
    }
    squash(diff1, diff2) {
        /* tslint:disable-next-line:strict-boolean-expressions */
        return (diff1 !== diff2) && undefined;
    }
}
exports.BooleanType = BooleanType;
