// kryo/types/custom.mjs
import { lazyProperties } from "../_helpers/lazy-properties";
import { createLazyOptionsError } from "../errors/lazy-options";
import { createNotImplementedError } from "../errors/not-implemented";
import { JSON_SERIALIZER } from "../json/index";
export const name = "custom";
export class CustomType {
    constructor(options) {
        this.name = name;
        // TODO: Remove once TS 2.7 is better supported by editors
        this.read = undefined;
        this.write = undefined;
        this.testError = undefined;
        this.equals = undefined;
        this.clone = undefined;
        this._options = options;
        if (typeof options !== "function") {
            this._applyOptions();
        }
        else {
            lazyProperties(this, this._applyOptions, ["read", "write", "testError", "equals", "clone"]);
        }
    }
    toJSON() {
        throw createNotImplementedError("CustomType#toJSON");
    }
    readTrustedJson(input) {
        return this.read(input, JSON_SERIALIZER);
    }
    readJson(input) {
        return this.read(input, JSON_SERIALIZER);
    }
    writeJson(value) {
        return this.write(value, JSON_SERIALIZER);
    }
    test(val) {
        return this.testError(val) === undefined;
    }
    _applyOptions() {
        if (this._options === undefined) {
            throw createLazyOptionsError(this);
        }
        const options = typeof this._options === "function" ? this._options() : this._options;
        Object.assign(this, {
            read: options.read,
            write: options.write,
            testError: options.testError,
            equals: options.equals,
            clone: options.clone,
        });
        Object.freeze(this);
    }
}
