// kryo/types/custom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lazy_properties_1 = require("../_helpers/lazy-properties");
const lazy_options_1 = require("../errors/lazy-options");
const not_implemented_1 = require("../errors/not-implemented");
const index_1 = require("../json/index");
exports.name = "custom";
class CustomType {
    constructor(options) {
        this.name = exports.name;
        // TODO: Remove once TS 2.7 is better supported by editors
        this.read = undefined;
        this.write = undefined;
        this.testError = undefined;
        this.equals = undefined;
        this.clone = undefined;
        this._options = options;
        if (typeof options !== "function") {
            this._applyOptions();
        }
        else {
            lazy_properties_1.lazyProperties(this, this._applyOptions, ["read", "write", "testError", "equals", "clone"]);
        }
    }
    toJSON() {
        throw not_implemented_1.createNotImplementedError("CustomType#toJSON");
    }
    readTrustedJson(input) {
        return this.read(input, index_1.JSON_SERIALIZER);
    }
    readJson(input) {
        return this.read(input, index_1.JSON_SERIALIZER);
    }
    writeJson(value) {
        return this.write(value, index_1.JSON_SERIALIZER);
    }
    test(val) {
        return this.testError(val) === undefined;
    }
    _applyOptions() {
        if (this._options === undefined) {
            throw lazy_options_1.createLazyOptionsError(this);
        }
        const options = typeof this._options === "function" ? this._options() : this._options;
        Object.assign(this, {
            read: options.read,
            write: options.write,
            testError: options.testError,
            equals: options.equals,
            clone: options.clone,
        });
        Object.freeze(this);
    }
}
exports.CustomType = CustomType;
