// kryo/types/json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const not_implemented_1 = require("../errors/not-implemented");
exports.name = "json";
class JsonType {
    constructor() {
        this.name = exports.name;
    }
    toJSON() {
        throw not_implemented_1.createNotImplementedError("ArrayType#toJSON");
    }
    readTrustedJson(input) {
        return input;
    }
    readJson(input) {
        return JSON.parse(JSON.stringify(input));
    }
    writeJson(val) {
        return JSON.parse(JSON.stringify(val));
    }
    testError(val) {
        try {
            JSON.parse(JSON.stringify(val));
            return undefined;
        }
        catch (err) {
            return err;
        }
    }
    test(val) {
        return this.testError(val) === undefined;
    }
    equals(val1, val2) {
        return JSON.stringify(val1) === JSON.stringify(val2);
    }
    clone(val) {
        return JSON.parse(JSON.stringify(val));
    }
    diff(oldVal, newVal) {
        throw not_implemented_1.createNotImplementedError("JsonType#diff");
    }
    patch(oldVal, diff) {
        throw not_implemented_1.createNotImplementedError("JsonType#patch");
    }
    reverseDiff(diff) {
        throw not_implemented_1.createNotImplementedError("JsonType#reverseDiff");
    }
    squash(diff1, diff2) {
        throw not_implemented_1.createNotImplementedError("JsonType#squash");
    }
}
exports.JsonType = JsonType;
