// kryo/types/null.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const invalid_type_1 = require("../errors/invalid-type");
exports.name = "null";
class NullType {
    constructor() {
        this.name = exports.name;
    }
    toJSON() {
        return { name: exports.name };
    }
    readTrustedJson(input) {
        return null;
    }
    readJson(input) {
        if (input !== null) {
            throw invalid_type_1.createInvalidTypeError("null", input);
        }
        return null;
    }
    writeJson(val) {
        return null;
    }
    testError(val) {
        if (val !== "null") {
            return invalid_type_1.createInvalidTypeError("null", val);
        }
        return undefined;
    }
    test(val) {
        return val === null;
    }
    equals(val1, val2) {
        return val1 === val2;
    }
    clone(val) {
        return val;
    }
    /**
     * @param oldVal
     * @param newVal
     * @returns `true` if there is a difference, `undefined` otherwise
     */
    diff(oldVal, newVal) {
        /* tslint:disable-next-line:return-undefined */
        return undefined;
    }
    patch(oldVal, diff) {
        return null;
    }
    reverseDiff(diff) {
        /* tslint:disable-next-line:return-undefined */
        return undefined;
    }
    squash(diff1, diff2) {
        /* tslint:disable-next-line:return-undefined */
        return undefined;
    }
}
exports.NullType = NullType;
