// kryo/types/try-union.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
}
Object.defineProperty(exports, "__esModule", { value: true });
const not_implemented_1 = require("../errors/not-implemented");
const union = __importStar(require("./union"));
exports.name = "try-union";
function toUnionOptions(options) {
    const variants = options.variants;
    const matcher = (value) => {
        for (const variant of variants) {
            if (variant.test(value)) {
                return variant;
            }
        }
        return undefined;
    };
    const readMatcher = (input, serializer) => {
        for (const variant of variants) {
            try {
                serializer.read(variant, input);
                return variant;
            }
            catch (err) {
                // Ignore error and try next variant
            }
        }
        return undefined;
    };
    return { variants: options.variants, matcher, readMatcher };
}
class TryUnionType extends union.UnionType {
    constructor(options) {
        super(typeof options === "function" ? () => toUnionOptions(options()) : toUnionOptions(options));
    }
    toJSON() {
        throw not_implemented_1.createNotImplementedError("TryUnionType#toJSON");
    }
    diff(oldVal, newVal) {
        throw not_implemented_1.createNotImplementedError("TryUnionType#diff");
    }
    patch(oldVal, diff) {
        throw not_implemented_1.createNotImplementedError("TryUnionType#patch");
    }
    reverseDiff(diff) {
        throw not_implemented_1.createNotImplementedError("TryUnionType#reverseDiff");
    }
    squash(diff1, diff2) {
        throw not_implemented_1.createNotImplementedError("TryUnionType#squash");
    }
}
exports.TryUnionType = TryUnionType;
