// kryo/types/white-list.mjs
import { Incident } from "incident";
import { lazyProperties } from "../_helpers/lazy-properties";
import { createLazyOptionsError } from "../errors/lazy-options";
import { createNotImplementedError } from "../errors/not-implemented";
import { JSON_SERIALIZER } from "../json";
export const name = "white-list";
export class WhiteListType {
    constructor(options) {
        this.name = name;
        // TODO: Remove once TS 2.7 is better supported by editors
        this.itemType = undefined;
        this.values = undefined;
        this._options = options;
        if (typeof options !== "function") {
            this._applyOptions();
        }
        else {
            lazyProperties(this, this._applyOptions, ["itemType", "values"]);
        }
    }
    static fromJSON(options) {
        throw createNotImplementedError("WhiteListType.fromJSON");
    }
    toJSON() {
        throw createNotImplementedError("TypedUnionType#toJSON");
    }
    readTrustedJson(val) {
        return JSON_SERIALIZER.readTrusted(this.itemType, val);
    }
    readJson(val) {
        const value = JSON_SERIALIZER.read(this.itemType, val);
        for (const allowed of this.values) {
            if (this.itemType.equals(value, allowed)) {
                return value;
            }
        }
        throw Incident("UnkownVariant", "Unknown variant");
    }
    writeJson(val) {
        return JSON_SERIALIZER.write(this.itemType, val);
    }
    testError(val) {
        const error = this.itemType.testError(val);
        if (error !== undefined) {
            return error;
        }
        for (const allowed of this.values) {
            if (this.itemType.equals(val, allowed)) {
                return undefined;
            }
        }
        return Incident("UnkownVariant", "Unknown variant");
    }
    test(val) {
        return this.testError(val) === undefined;
    }
    equals(val1, val2) {
        return this.itemType.equals(val1, val2);
    }
    clone(val) {
        return this.itemType.clone(val);
    }
    diff(oldVal, newVal) {
        return this.itemType.diff(oldVal, newVal);
    }
    patch(oldVal, diff) {
        return this.itemType.patch(oldVal, diff);
    }
    reverseDiff(diff) {
        return this.itemType.reverseDiff(diff);
    }
    squash(diff1, diff2) {
        return this.itemType.squash(diff1, diff2);
    }
    _applyOptions() {
        if (this._options === undefined) {
            throw createLazyOptionsError(this);
        }
        const options = typeof this._options === "function" ? this._options() : this._options;
        const itemType = options.itemType;
        const values = options.values;
        Object.assign(this, { itemType, values });
        Object.freeze(this);
    }
}
