// skype-http/connect.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
}
Object.defineProperty(exports, "__esModule", { value: true });
const incident_1 = require("incident");
const api = __importStar(require("./api"));
const context_1 = require("./interfaces/api/context");
const login_1 = require("./login");
const request_io_1 = require("./request-io");
/**
 * Authenticate the user and create a new API.
 *
 * @param options
 * @returns The Skype API for the provided user
 * @throws [[LoginError]]
 */
async function connect(options) {
    let apiContext;
    if (options.state !== undefined) {
        apiContext = context_1.Context.fromJson(options.state);
    }
    else if (options.credentials !== undefined) {
        apiContext = await login_1.login({
            io: request_io_1.requestIo,
            credentials: options.credentials,
            verbose: options.verbose,
        });
        if (options.verbose) {
            console.log("Obtained context trough authentication:");
            console.log({
                username: apiContext.username,
                skypeToken: apiContext.skypeToken,
                registrationToken: apiContext.registrationToken,
            });
        }
    }
    else {
        return Promise.reject(new incident_1.Incident("todo", "Connect must define `credentials`"));
    }
    return new api.Api(apiContext, request_io_1.requestIo);
}
exports.connect = connect;
