// skype-http/contacts/contacts.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
}
Object.defineProperty(exports, "__esModule", { value: true });
const incident_1 = require("incident");
const http_1 = require("../errors/http");
const get_contacts_1 = require("./api/get-contacts");
const get_invites_1 = require("./api/get-invites");
const contactsUrl = __importStar(require("./contacts-url"));
/**
 * @internal
 */
class ContactsService {
    constructor(httpIo) {
        this.httpIo = httpIo;
    }
    async getInvites(apiContext) {
        const url = contactsUrl.formatInvites(apiContext.username);
        const request = {
            uri: url,
            cookies: apiContext.cookies,
            headers: {
                "X-Skypetoken": apiContext.skypeToken.value,
            },
        };
        const response = await this.httpIo.get(request);
        if (response.statusCode !== 200) {
            http_1.UnexpectedHttpStatusError.create(response, new Set([200]), request);
        }
        let parsed;
        try {
            parsed = JSON.parse(response.body);
        }
        catch (err) {
            throw new incident_1.Incident(err, "UnexpectedResponseBody", { body: response.body });
        }
        let result;
        try {
            result = get_invites_1.$GetInvitesResult.readJson(parsed);
        }
        catch (err) {
            throw new incident_1.Incident(err, "UnexpectedResult", { body: parsed });
        }
        return result.inviteList;
    }
    async getContacts(apiContext) {
        return get_contacts_1.getContacts(this.httpIo, apiContext);
    }
}
exports.ContactsService = ContactsService;
