// skype-http/api/get-contact.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
}
Object.defineProperty(exports, "__esModule", { value: true });
const incident_1 = require("incident");
const apiUri = __importStar(require("../api-uri"));
const formatters_1 = require("../utils/formatters");
exports.VIRTUAL_CONTACTS = new Set(["concierge", "echo123"]);
async function getContact(io, apiContext, contactId) {
    if (exports.VIRTUAL_CONTACTS.has(contactId)) {
        // tslint:disable-next-line:max-line-length
        throw new Error(`${JSON.stringify(contactId)} is not a real contact, you cannot get data for ${JSON.stringify(contactId)}`);
    }
    const requestOptions = {
        uri: apiUri.userProfiles(),
        cookies: apiContext.cookies,
        form: { usernames: [contactId] },
        headers: {
            "X-Skypetoken": apiContext.skypeToken.value,
        },
    };
    const res = await io.post(requestOptions);
    if (res.statusCode !== 200) {
        return Promise.reject(new incident_1.Incident("net", "Unable to fetch contact"));
    }
    const body = formatters_1.formatSearchContact(JSON.parse(res.body)[0]);
    return body;
}
exports.getContact = getContact;
