// skype-http/api/get-conversations.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
}
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
}
Object.defineProperty(exports, "__esModule", { value: true });
const incident_1 = require("incident");
const lodash_1 = __importDefault(require("lodash"));
const messagesUri = __importStar(require("../messages-uri"));
const formatters_1 = require("../utils/formatters");
async function getConversations(io, apiContext) {
    const query = {
        startTime: "0",
        view: "msnp24Equivalent",
        targetType: "Passport|Skype|Lync|Thread",
    };
    const requestOptions = {
        uri: messagesUri.conversations(apiContext.registrationToken.host, messagesUri.DEFAULT_USER),
        cookies: apiContext.cookies,
        queryString: query,
        headers: {
            RegistrationToken: apiContext.registrationToken.raw,
        },
    };
    const res = await io.get(requestOptions);
    if (res.statusCode !== 200) {
        return Promise.reject(new incident_1.Incident("net", "Unable to fetch conversations"));
    }
    const body = JSON.parse(res.body);
    return lodash_1.default.map(body.conversations, formatters_1.formatConversation);
}
exports.getConversations = getConversations;
