// skype-http/api/send-message.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
}
Object.defineProperty(exports, "__esModule", { value: true });
const incident_1 = require("incident");
const messagesUri = __importStar(require("../messages-uri"));
const utils_1 = require("../utils");
async function sendMessage(io, apiContext, message, conversationId) {
    const query = {
        clientmessageid: String(utils_1.getCurrentTime() + Math.floor(10000 * Math.random())),
        content: String(message.textContent),
        messagetype: "RichText",
        contenttype: "text",
    };
    const requestOptions = {
        uri: messagesUri.messages(apiContext.registrationToken.host, messagesUri.DEFAULT_USER, conversationId),
        cookies: apiContext.cookies,
        body: JSON.stringify(query),
        headers: {
            RegistrationToken: apiContext.registrationToken.raw,
        },
    };
    const res = await io.post(requestOptions);
    console.log(JSON.stringify(res, null, 2));
    if (res.statusCode !== 201) {
        return Promise.reject(new incident_1.Incident("send-message", "Received wrong return code"));
    }
    const parsed = messagesUri.parseMessage(res.headers["location"]);
    const body = JSON.parse(res.body);
    return {
        clientMessageId: query.clientmessageid,
        arrivalTime: body.OriginalArrivalTime,
        textContent: query.content,
        MessageId: parsed.message,
    };
}
exports.sendMessage = sendMessage;
