// skype-http/api/set-status.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
}
Object.defineProperty(exports, "__esModule", { value: true });
const incident_1 = require("incident");
const messagesUri = __importStar(require("../messages-uri"));
async function setStatus(io, apiContext, status) {
    const requestBody = {
        status,
    };
    const requestOptions = {
        uri: messagesUri.userMessagingService(apiContext.registrationToken.host),
        cookies: apiContext.cookies,
        body: JSON.stringify(requestBody),
        headers: {
            RegistrationToken: apiContext.registrationToken.raw,
        },
    };
    const res = await io.put(requestOptions);
    if (res.statusCode !== 200) {
        return Promise.reject(new incident_1.Incident("send-message", "Received wrong return code"));
    }
}
exports.setStatus = setStatus;
